base/genericmemory.jl

```diff
@@ -0,0 +1,273 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+## genericmemory.jl: Managed Memory
+
+"""
+    GenericMemory{isatomic::Bool, addrspace::Int, T} <: AbstractVector{T}
+
+One-dimensional dense array with elements of type `T`.
+"""
+GenericMemory
+"""
+    Memory{T} == GenericMemory{:not_atomic, 0, T}
+
+One-dimensional dense array with elements of type `T`.
+"""
+Memory
+
+## Basic functions ##
+
+using Core: memoryrefoffset, memoryref_isassigned # import more functions which were not essential
+
+size(a::GenericMemory, d::Int) =
+    d < 1 ? error("dimension out of range") :
+    d == 1 ? length(a) :
+    1
+size(a::GenericMemory, d::Integer) =  size(a, convert(d, Int))
+size(a::GenericMemory) = (length(a),)
+
+pointer(mem::GenericMemory, i::Int) = (@_propagate_inbounds_meta; unsafe_convert(Ptr{Cvoid}, GenericMemoryRef(mem, i))) # boundschecked, even for i==1
+pointer(mem::GenericMemoryRef) = unsafe_convert(Ptr{Cvoid}, mem) # no bounds check, even for empty array
+
+_unsetindex!(A::Memory, i::Int) =  (@_propagate_inbounds_meta; _unsetindex!(GenericMemoryRef(A, i)); A)
+function _unsetindex!(A::MemoryRef{T}) where T
+    @_terminates_locally_meta
+    @_propagate_inbounds_meta
+    @inline
+    @boundscheck GenericMemoryRef(A, 1)
+    mem = A.mem
+    MemT = typeof(mem)
+    arrayelem = datatype_arrayelem(MemT)
+    elsz = datatype_layoutsize(MemT)
+    isboxed = 1; isunion = 2
+    t = @_gc_preserve_begin mem
+    p = Ptr{Ptr{Cvoid}}(@inbounds pointer(A))
+    if arrayelem == isboxed
+        Intrinsics.atomic_pointerset(p, C_NULL, :monotonic)
+    elseif arrayelem != isunion
+        if !datatype_pointerfree(T::DataType)
+            for j = 1:Core.sizeof(Ptr{Cvoid}):elsz
+                Intrinsics.atomic_pointerset(p + j - 1, C_NULL, :monotonic)
+            end
+        end
+    end
+    @_gc_preserve_end t
+    return A
+end
+
+elsize(@nospecialize _::Type{A}) where {T,A<:GenericMemory{<:Any,<:Any,T}} = aligned_sizeof(T)
+sizeof(a::GenericMemory) = Core.sizeof(a)
+
+# multi arg case will be overwritten later. This is needed for bootstraping
+function isassigned(a::Memory, i::Int)
+    @inline
+    @boundscheck 1 <= i <= length(a) || return false
+    # @boundscheck (i % UInt) - 1 < length(A) || return false
+    return @inbounds memoryref_isassigned(GenericMemoryRef(a, i), :not_atomic, false)
+end
```

Comment:
As far as I can tell the C methods backing `memoryref_isassigned` aren't used in "src", so we might be able to move this completely to Julia code and clean up some of the the source code. Here's what I was thinking.

```julia
function isassigned(mem::Memory, i::Int)
    @inline
    @boundscheck 1 <= i <= length(mem) || return false
    memT = typeof(mem)
    arrayelem = datatype_arrayelem(memT)
    elsz = datatype_layoutsize(memT)
    # might need to be mad
    firstptr = unsafe_load(unsafe_convert(Ptr{DataTypeLayout}, memT.layout)).firstptr % Int
    isboxed = 1; isunion = 2
    if arrayelem == isboxed
        p = Ptr{Ptr{Cvoid}}(@inbounds pointer(mem))
        t = @_gc_preserve_begin mem
        out = Intrinsics.atomic_pointerref(p, :monotonic) != C_NULL
        @_gc_preserve_end t
    elseif arrayelem != isunion && firstptr >= 0
        p = Ptr{Ptr{Cvoid}}(@inbounds pointer(mem)) + fieldoffset(eltype(memT), firstptr + 1)
        t = @_gc_preserve_begin mem
        out = Intrinsics.atomic_pointerref(p, :monotonic) != C_NULL
        @_gc_preserve_end t
    else
        out = true
    end
    return out
end
```

Reply:
We could, but that code is less correct than the code it replaces, so no, that is not a good implementation

Reply:
But could we use a Julia implementation instead? This was just my first quick attempt, so if we can still get performance and correctness in Julia it seems like a big win to get rid of all that extra source code.

Reply:
the fact that this necessarily involves casting to `Ptr{Ptr{Cvoid}}` implies to me that this should probably stay on the C side. Keeping the Julia side "normal" is nice.

Reply:
It's not the only place in Julia code with `Memory` that this is handled, but I do see the merit in being cautious nonetheless since we don't have a fully public API for this kind of pointer.

Reply:
This probably should be an intrinsic